### ISSUE TITLE
triangulate_points returns a point at infinity as a finite, unflagged point whose position is set by roundoff

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4865]
`triangulate_points` finds the homogeneous DLT solution `(X, w)` and passes it to `convert_points_from_homogeneous`. That function divides by `w` when `|w| > 1e-8` and otherwise returns `X` unchanged. A correspondence whose two back-projected rays are parallel is the image of a direction `(x, y, z, 0)`, a point at infinity such as a distant landmark or a star, so its exact `w` is `0` and the solver's `w` is roundoff. With every solver (`svd`, `eigh`, `cofactor`), the output is a finite Euclidean point with nothing to mark it. Its magnitude depends on which side of `1e-8` the roundoff lands and on the solver: the unit direction (norm `1`, or `2` for `cofactor`), or a point `2.8e6` to `2.9e7` away. Its sign is arbitrary as well, and in float32 most of these points come back behind the first camera (`z < 0`). OpenCV's `triangulatePoints` returns homogeneous `(4, N)` output with `|w| / |X|` at roundoff level, so a caller can detect the point at infinity and keep its direction.

## Repro

Two constructions on the same camera pair: A projects three unit directions, and B projects the unnormalised direction `(0.3, -0.2, 1.0)`. Both are run in float32 and float64. The two-line spy records the `w` that reaches the dehomogenisation, to show where the magnitude comes from.

```python
import torch
import kornia.geometry.epipolar.triangulation as tri
from kornia.geometry.conversions import axis_angle_to_rotation_matrix
from kornia.geometry.epipolar import projection_from_KRt, triangulate_points

# record the homogeneous w that reaches the final dehomogenisation
w = []
_c = tri.convert_points_from_homogeneous
tri.convert_points_from_homogeneous = lambda p: (w.append(p[..., 3].abs()), _c(p))[1]


def project(P, Xh):  # Xh: (N, 4) homogeneous world points
    x = Xh @ P.T
    return x[:, :2] / x[:, 2:]


for D in [torch.float32, torch.float64]:
    K1 = torch.tensor([[800.0, 0.0, 320.0], [0.0, 760.0, 200.0], [0.0, 0.0, 1.0]], dtype=D)
    K2 = torch.tensor([[700.0, 0.0, 300.0], [0.0, 740.0, 240.0], [0.0, 0.0, 1.0]], dtype=D)
    R = axis_angle_to_rotation_matrix(torch.tensor([[0.1, -0.2, 0.05]], dtype=D))[0]
    t = torch.tensor([[0.5], [0.1], [0.02]], dtype=D)
    P1 = projection_from_KRt(K1, torch.eye(3, dtype=D), torch.zeros(3, 1, dtype=D))
    P2 = projection_from_KRt(K2, R, t)
    d = torch.tensor([[0.1, -0.05, 1.0], [-0.2, 0.1, 1.0], [0.05, 0.2, 1.0]], dtype=D)
    cases = {"A: three unit directions": d / d.norm(dim=1, keepdim=True),
             "B: direction (0.3, -0.2, 1.0)": torch.tensor([[0.3, -0.2, 1.0]], dtype=D)}
    for name, dirs in cases.items():
        Xh = torch.cat([dirs, torch.zeros(len(dirs), 1, dtype=D)], 1)  # points at infinity, w = 0
        x1, x2 = project(P1, Xh), project(P2, Xh)
        print(D, name)
        for solver in ["svd", "eigh", "cofactor"]:
            X = triangulate_points(P1, P2, x1, x2, solver=solver)
            print(f"  {solver:8s} norm {[f'{v:.4g}' for v in X.norm(dim=1).tolist()]}  "
                  f"z {[f'{v:.3g}' for v in X[:, 2].tolist()]}  |w| {[f'{v:.1e}' for v in w[-1].tolist()]}  "
                  f"finite {bool(torch.isfinite(X).all())}")
```

```
torch.float32 A: three unit directions
  svd      norm ['1', '2.948e+07', '1.871e+07']  z ['0.994', '-2.88e+07', '-1.83e+07']  |w| ['9.1e-10', '3.4e-08', '5.3e-08']  finite True
  eigh     norm ['1', '2.948e+07', '1.871e+07']  z ['0.994', '-2.88e+07', '-1.83e+07']  |w| ['9.1e-10', '3.4e-08', '5.3e-08']  finite True
  cofactor norm ['2.682e+07', '8.584e+06', '2.824e+06']  z ['2.67e+07', '-8.38e+06', '-2.77e+06']  |w| ['7.5e-08', '2.3e-07', '7.1e-07']  finite True
torch.float32 B: direction (0.3, -0.2, 1.0)
  svd      norm ['1.025e+07']  z ['-9.64e+06']  |w| ['9.8e-08']  finite True
  eigh     norm ['1.025e+07']  z ['-9.64e+06']  |w| ['9.8e-08']  finite True
  cofactor norm ['1.877e+07']  z ['-1.77e+07']  |w| ['1.1e-07']  finite True
torch.float64 A: three unit directions
  svd      norm ['1', '1', '1']  z ['0.994', '0.976', '-0.979']  |w| ['2.0e-16', '4.7e-16', '3.2e-16']  finite True
  eigh     norm ['1', '1', '1']  z ['0.994', '0.976', '0.979']  |w| ['1.4e-15', '1.3e-15', '9.2e-16']  finite True
  cofactor norm ['2', '2', '2']  z ['-1.99', '-1.95', '-1.96']  |w| ['1.0e-16', '2.8e-16', '9.4e-16']  finite True
torch.float64 B: direction (0.3, -0.2, 1.0)
  svd      norm ['1']  z ['0.941']  |w| ['3.2e-16']  finite True
  eigh     norm ['1']  z ['-0.941']  |w| ['2.4e-16']  finite True
  cofactor norm ['2']  z ['-1.88']  |w| ['6.7e-16']  finite True
```

In float64 the roundoff `w` is below `1e-8`, so the numerator comes back unchanged: a unit-distance point (`2` for `cofactor`, which sums two unit vectors). In float32 the roundoff `w` is mostly above `1e-8`, so the output is `X / w`, millions of units away, and its sign follows the sign of the roundoff (`z < 0` in 9 of the 12 float32 rows). The first float32 `svd`/`eigh` point, with `|w| = 9.1e-10`, falls into the unit-distance branch. Nothing in the output tells these rows apart from real points at those distances.

OpenCV on the same constructions (run separately with `uv run --no-project --with opencv-python-headless --with numpy python`):

```python
import cv2
import numpy as np

K1 = np.array([[800.0, 0.0, 320.0], [0.0, 760.0, 200.0], [0.0, 0.0, 1.0]])
K2 = np.array([[700.0, 0.0, 300.0], [0.0, 740.0, 240.0], [0.0, 0.0, 1.0]])
R = cv2.Rodrigues(np.array([0.1, -0.2, 0.05]))[0]
t = np.array([[0.5], [0.1], [0.02]])
P1 = K1 @ np.hstack([np.eye(3), np.zeros((3, 1))])
P2 = K2 @ np.hstack([R, t])
d = np.array([[0.1, -0.05, 1.0], [-0.2, 0.1, 1.0], [0.05, 0.2, 1.0]])
cases = {"A": d / np.linalg.norm(d, axis=1, keepdims=True), "B": np.array([[0.3, -0.2, 1.0]])}
for dt in [np.float64, np.float32]:
    for name, dirs in cases.items():
        Xh = np.hstack([dirs, np.zeros((len(dirs), 1))])
        x1 = (P1 @ Xh.T); x1 = x1[:2] / x1[2]
        x2 = (P2 @ Xh.T); x2 = x2[:2] / x2[2]
        out = cv2.triangulatePoints(P1.astype(dt), P2.astype(dt), x1.astype(dt), x2.astype(dt))  # (4, N) homogeneous
        print(dt.__name__, name, "out dtype", out.dtype, "|w| / |X|:", [f"{v:.1e}" for v in (np.abs(out[3]) / np.linalg.norm(out, axis=0)).tolist()])
```

```
float64 A out dtype float64 |w| / |X|: ['6.9e-17', '6.3e-17', '6.1e-17']
float64 B out dtype float64 |w| / |X|: ['1.3e-17']
float32 A out dtype float32 |w| / |X|: ['2.0e-08', '5.1e-09', '2.9e-08']
float32 B out dtype float32 |w| / |X|: ['6.9e-08']
```

kornia main 7ddf731bd, torch 2.14.0 (and 2.5.1: same, byte-identical output), CPU float32 and float64, macOS arm64; OpenCV 5.0.0.

## Classification

Silent wrong result. Idiom 4 (degenerate or singular input silently producing garbage instead of raising): a correspondence with parallel rays has no finite solution, and the Euclidean-only return type leaves no way to say so. The output distance is not a property of the scene. It comes from roundoff and from the `1e-8` branch in `convert_points_from_homogeneous`.

## Disposition

MERGE IN WINDOW — behaviour change: correspondences with parallel rays stop coming back as ordinary finite points. There are two ways to do it: return the homogeneous `(*, N, 4)` solution, either as an option or as the default, or mark those rows invalid. The choice should follow the validity-mask decision in #4555, which is about the same problem for projections and depth warps. Every caller that relies on today's always-finite Euclidean output is affected.
Found by the batch-8 conventions audit (two-view geometry); a `test_wart_*` pin in the docs PR flips when the fix lands.

Related: #4863 (same function: solver="cofactor" returns NaN in float16).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
